### ISSUE TITLE
[Android 12] Fatal Exception: android.app.ForegroundServiceStartNotAllowedException: startForegroundService() not allowed due to mAllowStartForeground false

### ISSUE THREAD

**Issue by KaranShah07** [transistorsoft/react-native-background-geolocation#1444]
Hello, we are getting a lot of Crashlytics issues reported for our app on the android 12 devices. we do use background geolocation to monitor geofences only.

## Your Environment
* Platform: Android
* OS version: 12
* Device manufacturer / model: Galaxy Note10+, Galaxy S21 Ultra 5G, Pixel 6 Pro, Pixel 4, Pixel 4a
* React Native version (`react-native -v`): 0.64.2
* Plugin config
```javascript <!-- Syntax-highlighting:  paste your code below -->
const config = {
      enableHeadless: true,
      geofenceModeHighAccuracy: true,
      desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
      distanceFilter: 10,
      geofenceInitialTriggerEntry: false,
      disableMotionActivityUpdates: true,
      stopOnTerminate: false,
      url,
    };

    await BackgroundGeolocation.ready(config);

    BackgroundGeolocation.startGeofences();
```

## Expected Behavior
App should never crash while attempting to start foreground service on android and geofences should work fine for all OS versions and devices.

## Actual Behavior
App is crashing when plugin tries to start foreground service on android 12 devices.

## Context
Crash Report:
Fatal Exception: android.app.ForegroundServiceStartNotAllowedException: startForegroundService() not allowed due to mAllowStartForeground false: service com.transistorsoft.locationmanager.service.TrackingService
       at android.app.ForegroundServiceStartNotAllowedException$1.createFromParcel(ForegroundServiceStartNotAllowedException.java:54)
       at android.app.ForegroundServiceStartNotAllowedException$1.createFromParcel(ForegroundServiceStartNotAllowedException.java:50)
       at android.os.Parcel.readParcelable(Parcel.java:3345)
       at android.os.Parcel.createExceptionOrNull(Parcel.java:2432)
       at android.os.Parcel.createException(Parcel.java:2421)
       at android.os.Parcel.readException(Parcel.java:2404)
       at android.os.Parcel.readException(Parcel.java:2346)
       at android.app.IActivityManager$Stub$Proxy.startService(IActivityManager.java:6914)
       at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1926)
       at android.app.ContextImpl.startForegroundService(ContextImpl.java:1892)
       at android.content.ContextWrapper.startForegroundService(ContextWrapper.java:796)
       at com.transistorsoft.locationmanager.service.AbstractService$a.run(:10)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at android.app.ActivityThread.main(ActivityThread.java:8641)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:567)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1133)

Caused by android.os.RemoteException: Remote stack trace:
	at com.android.server.am.ActiveServices.startServiceLocked(ActiveServices.java:771)
	at com.android.server.am.ActiveServices.startServiceLocked(ActiveServices.java:679)
	at com.android.server.am.ActivityManagerService.startService(ActivityManagerService.java:13987)
	at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:2943)
	at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:3030)

**Comment by spatialvlad** [comment]
It looks like I'm having an issue with similar behavior on Android 11, Samsung Galaxy S21 Ultra. Use-case is similar - Geolocation only tracking.

The first time the plugin starts, gets needed permissions, but later crashes.
 
Attempts on next starts to call ready lead to a crash.

This line still works:
`const status = await BackgroundGeolocation.getProviderState();`

and gives the result:
```
status: {
  accuracyAuthorization: 0,
  enabled: true,
  gps: true,
  network: true,
  status: 3,
}
```

After calling this line I get a crash:

```
  await BackgroundGeolocation.ready({
    desiredAccuracy:
      BackgroundGeolocation.DESIRED_ACCURACY_MEDIUM,
    distanceFilter: 50,
    debug: true,
    logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
    enableHeadless: true,
    geofenceInitialTriggerEntry: false,
    locationAuthorizationRequest: 'Always',
    stopOnTerminate: false,
    startOnBoot: true,
    showsBackgroundLocationIndicator: false,
  });
```

Relevant log output:
```
2022-02-12 11:08:54.090 10523-10523/com.myapp E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.myapp, PID: 10523
    android.app.RemoteServiceException: Bad notification for startForeground
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2258)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:246)
        at android.app.ActivityThread.main(ActivityThread.java:8633)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
2022-02-12 11:08:54.092 10523-10523/com.myapp E/TSLocationManager: [c.t.l.a.BackgroundGeolocation$x0 uncaughtException] 
      ‼️  Uncaught Exception: Bad notification for startForeground
    {"activityRecognitionInterval":10000,"allowIdenticalLocations":false,"authorization":{},"autoSync":true,"autoSyncThreshold":0,"backgroundPermissionRationale":{"title":"Allow {applicationName} to access this device's location even when closed or not in use?","message":"[CHANGEME] This app collects location data for FEATURE X and FEATURE Y.","positiveAction":"Change to \"{backgroundPermissionOptionLabel}\"","negativeAction":""},"batchSync":false,"configUrl":"","debug":true,"deferTime":0,"desiredAccuracy":10,"desiredOdometerAccuracy":100,"disableAutoSyncOnCellular":false,"disableElasticity":false,"disableLocationAuthorizationAlert":false,"disableMotionActivityUpdates":false,"disableStopDetection":false,"distanceFilter":50,"elasticityMultiplier":1,"enableHeadless":true,"enableTimestampMeta":false,"extras":{},"fastestLocationUpdateInterval":-1,"foregroundService":true,"geofenceInitialTriggerEntry":false,"geofenceModeHighAccuracy":false,"geofenceProximityRadius":1000,"geofenceTemplate":"","headers":{},"headlessJobService":"com.transistorsoft.rnbackgroundgeolocation.HeadlessTask","heartbeatInterval":-1,"httpRootProperty":"location","httpTimeout":60000,"isMoving":false,"locationAuthorizationRequest":"Always","locationTemplate":"","locationTimeout":60,"locationUpdateInterval":1000,"locationsOrderDirection":"ASC","logLevel":5,"logMaxDays":3,"maxBatchSize":-1,"maxDaysToPersist":1,"maxRecordsToPersist":-1,"method":"POST","minimumActivityRecognitionConfidence":75,"motionTriggerDelay":0,"notification":{"layout":"","title":"","text":"Location Service activated","color":"","channelName":"TSLocationManager","smallIcon":"","largeIcon":"","priority":0,"sticky":false,"strings":{},"actions":[]},"params":{},"persist":true,"persistMode":2,"schedule":[],"scheduleUseAlarmManager":false,"speedJumpFilter":300,"startOnBoot":true,"stationaryRadius":25,"stopAfterElapsedMinutes":0,"stopOnStationary":false,"stopOnTerminate":false,"stopTimeout":5,"triggerActivities":"in_vehicle, on_bicycle, on_foot, running, walking","url":"","useSignificantChangesOnly":false,"enabled":true,"schedulerEnabled":false,"trackingMode":0,"odometer":0,"isFirstBoot":false,"didLaunchInBackground":false,"didDeviceReboot":false}
    android.app.RemoteServiceException: Bad notification for startForeground
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2258)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:246)
        at android.app.ActivityThread.main(ActivityThread.java:8633)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

**Comment by christocracy** [comment]
Can you create for me a simple Hello World app that reproduces this?

**Comment by spatialvlad** [comment]
Ok, I will do it by the end of the day today (US time).

**Comment by christocracy** [comment]
Publish it to a public GitHub repo for me

**Comment by spatialvlad** [comment]
I checked the lines, I did have a different/custom `android:icon` name, but the custom one was present.
I changed the name both in AndroidManifest and in the res folder, but unfortunately, it keeps failing at the same place.
Thanks for the suggestion though!

**Comment by christocracy** [comment]
In Android Studio, try generating a new icon using the built-in clip art.  Re-configure your AndroidManifest to point to that new icon (both icon attributes)

**Comment by spatialvlad** [comment]
@christocracy , I tired your last suggestion, it did not help, unfortunately.

**Comment by christocracy** [comment]
What version of the plugin is in your package.json?

**Comment by spatialvlad** [comment]
```
    "react-native": "0.66.4",
    "react-native-background-fetch": "^4.0.5",
    "react-native-background-geolocation": "^4.4.4",
```

**Comment by christocracy** [comment]
> Bad notification for startForeground

Actually, your crash has nothing to do with the OP issue.

> Fatal Exception: android.app.ForegroundServiceStartNotAllowedException: startForegroundService() not allowed due to mAllowStartForeground false

Post a new issue.  Also have you searched existing issues?  I’ve heard of this before.

**Comment by KaranShah77** [comment]
Our team doesn’t have any exact repo steps yet, but a lot of crashes are getting reported after we have updated plugin.

For our app we are using these plugin versions
`react-native-background-geolocation: 4.4.4`
`react-native-background-fetch: 4.0.4`

I also have gone through CHANGELOG, is there any specific change that you would like me to check on app?

**Comment by KaranShah07** [comment]
Our team doesn’t have any exact repo steps yet, but a lot of crashes are getting reported after we have updated plugin.

For our app we are using these plugin versions
"react-native-background-fetch": "^4.0.4",
    "react-native-background-geolocation": "^4.4.4",

I also have gone through CHANGELOG, is there any specific change that you would like me to check on app?

**Comment by christocracy** [comment]
Show me the date(s) of these crashes you're seeing.

**Comment by christocracy** [comment]
Sorry, I see it in the logs.

>2022-02-12 11:08:54.090

Are you *sure* those crashes correspond to a version of your app where you're using `v4.4.4`?

**Comment by KaranShah07** [comment]
The first crash was reported on Jan 11, 2022. After that we are getting minimum of 3 crashes everyday.
Yes I'm sure we started getting those crashes after we have released our app with the plugin updation changes. In our last release we have included changes for updating plugin version to make it workable on android 12.

**Comment by ozberkctn** [comment]
We are facing this issue on Android 12 devices. We can not produce but we have crash reports on Firebase. @christocracy 

Plugin version: 4.4.4

Our initial SDK code
`BackgroundGeolocation.ready({
        desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
        stationaryRadius: 50,
        distanceFilter: this.props.user.tracking_distance,
        startOnBoot: true,
        stopOnTerminate: false,
        debug: false, // <-- enable this hear sounds for background-geolocation life-cycle.
        logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
        enableHeadless: true,
        stopTimeout: 5,
        notification: {
          title: 'Location Tracking',
          text: 'ACTIVE',
        },
        url: `${MAIN_URL}/user/user-tracker`,
        headers: {
          'jm-token': jmDeviceToken,
          'jm-version': version,
          'jm-device': deviceNumber.toString(),
          'jm-user-token': this.props.token,
        },
        params: {
          platform: Platform.OS,
        },
      }).then((state) => {
        resolve(state);
      });
    });`

I can provide you with any information to solve that issue anytime. I am looking forward to hearing from you.

**Comment by christocracy** [comment]
This is caused by attempting to call `.start()` or `.getCurrentPosition()` while your app is in the background where the plugin is currently disabled (ie: `State.enabled == false`).  It is due to new [Android 12 restrictions on foreground-services](https://developer.android.com/about/versions/12/behavior-changes-12#foreground-service-launch-restrictions)

**Comment by christocracy** [comment]
I have [found a solution](https://github.com/transistorsoft/flutter_background_geolocation/issues/703#issuecomment-1116160289) to calling `.start()` and `.getCurrentPosition()` when the plugin is `State.enabled == false` and in the background.

It will be merged here for RN soon.

**Comment by christocracy** [comment]
Ok, guarding for this Android 12 exception is now merged to the private repo.

**Comment by ozberkctn** [comment]
Hi @christocracy I really appreciate for your quick solution.

**Comment by ZerozAndOnes** [comment]
This issue still exists on the latest version of [4.7.1](https://github.com/transistorsoft/react-native-background-geolocation/issues/1452#issuecomment-1120262386).
With the relevant [fix](https://github.com/transistorsoft/react-native-background-geolocation/pull/1480), the occurrences of this have gone done, but the issue still exists at least with `.start()`, it seems to be fixed for `.getCurrentPosition()`

I was able to reproduce this by doing the RN equivalent of the repro steps for [flutter](https://github.com/transistorsoft/flutter_background_geolocation/issues/703)

### Repro steps
1. Set [`logLevel`](https://transistorsoft.github.io/react-native-background-geolocation/interfaces/config.html#loglevel) to be `LOG_LEVEL_VERBOSE`
2. Give the max location permission possible to the app
3. With your app in the foreground, call `BackgroundGeolocation.stop()`.
4. Put app in background.
5. Wait 20s.
6. Call `BackgroundGeolocation.start()` when app is in background
We did this through a HeadlessJS task (We created an intent for `android.net.conn.CONNECTIVITY_CHANGE` and a service which would be run when this intent was received, the service then ran `BackgroundGeolocation.start()` after 20s

- If before calling we check for state.enabled, then the error doesn't happen as it's false.
- If instead of `BackgroundGeolocation.start()` we do `BackgroundGeolocation.getCurrentPosition()` in step 5, then the error happen.

### Additional information:

- The issue happens on android 12 only and when the app is 100% in the background
- In our code, before calling `BackgroundGeolocation.start()` we do check for `State.enabled`, thus, there's possibly another way to reproduce this too

### Relevant Logs
- Shared through email

**Comment by christocracy** [comment]
@ZerozAndOnes I've reproduced this.  I added a little bit of code to immediately launch a foreground-service when `.start()` is called but it seems under these conditions that OS will only allow the foreground-service to be launched specifically due to a `location` API request.

I can work around this.

**Comment by christocracy** [comment]
Ok, turns out it was just a blunder on my part.  The SDK has a general purpose "one-shot" timer mechanism built into it which uses `JobScheduler` by default.  I neglected to pass the 3rd optional param `forceAlarmManager: true`.  

<img width="1030" alt="Screen Shot 2022-05-27 at 9 32 42 AM" src="https://user-images.githubusercontent.com/14999/170710264-e4a1264e-7695-4a47-9cba-604e59f68298.png">

Only `AlarmManager`-based events are granted *carte blanche* by the OS to launch foreground-services.

@ZerozAndOnes, @ozberkctn Please test branch `#release-4.7.2` just pushed.

```bash
yarn add https://github.com/transistorsoft/react-native-background-geolocation.git#release-4.7.2
```

**Comment by christocracy** [comment]
You will know you have `4.7.2` installed when you see this build number in the logs:

```
TSLocationManager: ╔═════════════════════════════════════════════
TSLocationManager: ║ TSLocationManager version: 3.1.44 (400)
TSLocationManager: ╠═════════════════════════════════════════════
```

**Comment by christocracy** [comment]
The `getCurrentPosition` method was *correctly* using `forceAlarmManager: true`.  When initially developing this with my test case, I initially implemented with `start` method.  It was only after moving onto `getCurrentPosition` that I discovered that the OS provides a little leeway on background time (about 10s) to launch FG services.  After I increasing the `Timer` to `15 seconds`, the exception was raised and I remembered to use `forceAlarmManager: true` on the ONESHOT.  

I just didn't circle back to `.start()`.

**Comment by ZerozAndOnes** [comment]
Thanks Chris, will test out 4.7.2, and will report back if there are any further crashes.
Fingers crossed, hopefully, there had be none 🤞 
Thanks for the quick fix 🤜 🤛

**Comment by ZerozAndOnes** [comment]
@christocracy Unfortunately, the error is still there. I was to reproduce it mostly using the same reproduction, though, sometimes it was working fine with these steps the first time, but repeating them again would result in the error:
- With your app in the foreground, call BackgroundGeolocation.stop().
- Put app in background.
- Wait 20s.
- Call BackgroundGeolocation.start() when app is in background


- Shared the logs in email again.

**Comment by christocracy** [comment]
Show me your code.

**Comment by ZerozAndOnes** [comment]
Shared, thanks!

**Comment by ZerozAndOnes** [comment]
Not in succession, but rather just repeating the above repro steps.
Since it seems like it isn't reproducible on your end, will set up log traces for `BackgroundGeolocation` function calls and will do a partial release to get more info on this. 
Personally, I haven't been able to reproduce it under normal usage.

---
P.S: I totally agree on the proprietary point, however, since I am working on a proprietary product, couldn't risk sharing anything publicly by mistake. 
Will make sure to share the root cause, if it turns out to be an issue from our end.

**Comment by christocracy** [comment]
I believe that somehow your code is going this while plugin is in background and disabled:

```javascript
BackgroundGeolocation.start();
.
. a few milliseconds later:
.
BackgroundGeolocation.start();
```

I will guard against this.

I can see it in your logs:

```
2022-05-30 19:06:03.883 23965-24189/com.domain.project.dev I/TSLocationManager: [c.t.l.s.TSScheduleManager oneShot] 
      ⏰ Scheduled OneShot: START in 1ms (jobID: 79219778).  <--- GOOD, ONESHOT fired.
2022-05-30 19:06:07.679 23965-24189/com.domain.project.dev D/TSLocationManager: [c.t.locationmanager.util.c g] 
      ℹ️  LocationAuthorization: Permission granted
2022-05-30 19:06:07.681 23965-24456/com.domain.project.dev I/TSLocationManager: - Enable: false → true, trackingMode: 1.  // <--------- HERE is .start() begin called a 2nd time
```

**Comment by christocracy** [comment]
It's already in the private repo, yes.

**Comment by christocracy** [comment]
4.7.2 branch is not yet *released*.  It's not tagged yet.

**Comment by christocracy** [comment]
I've got another release-candidate in the *private repo* [release-4.8.0](https://github.com/transistorsoft/react-native-background-geolocation-android/tree/release-4.8.0) brewing with significant *Android 12* improvements.

**Comment by christocracy** [comment]
@ZerozAndOnes [`release-4.8.0`](https://github.com/transistorsoft/react-native-background-geolocation-android/tree/release-4.8.0) branch in the private repo implements a *much* more generalized solution for the `.getCurrentPosition()` and `.start()` problem in the background.

The SDK no longer needs to be concerned specifically with any particular method &mdash; it now simply does a `catch (ForegroundServiceStartNotAllowedException e) { ... }` and does the `AlarmManager` *oneShot* event there, for **every use-case**.

## 4.8.0 &mdash; Unreleased
* [Android] __Android 12__:  Guard `Context.startForegroundService` with `try / catch`: the plugin will now catch exception `ForegroundServiceStartNotAllowedException` and automatically retry with an `AlarmManager` *oneShot* event.
* [Android] __Android 12__: Refactor foreground-service management for Android 12:  A way has been found to restore the traditional behaviour of foreground-services, allowing them to stop when no longer required (eg: where the plugin is in the stationary state).
* [Android] Refactor application life-cycle management.  Remove deprecated permission `android.permission.GET_TASKS` traditionally used for detecting when the app has been terminated.  The new life-cycle mgmt system can detect Android headless-mode in a much more elegant manner.
* [Android] Better handling for `WhenInUse` behaviour:  The plugin will not allow `.changePace(true)` to be executed when the app is in the background (since Android forbids location-services to initiated in the background with `WhenInUse`).
* [Android] Refactor `useSignificantChangesOnly` behaviour.  Will use a default `motionTriggerDelay` with minimum 60000ms, minimum `distanceFilter: 250` and enforced `stopTimeout: 20`.
* [iOS] iOS 15 has finally implemented *Mock Location Detection*.  `location.mock` will now be present for iOS when the location is mocked, just like Android.

**Comment by ZerozAndOnes** [comment]
@christocracy Yes, nothing came up during our internal tests and seems good to go. 
Have a release waiting Google's approval, will get to know properly then 🤞 
Will keep you posted.

**Comment by ZerozAndOnes** [comment]
This issue hasn't come up in our latest release, we can close this issue and the following ones in the private repo:

- [1181](https://github.com/transistorsoft/react-native-background-geolocation-android/issues/1181)
- [1137](https://github.com/transistorsoft/react-native-background-geolocation-android/issues/1137)
